services/ingestion/sealing_verifier.go

```diff
@@ -0,0 +1,256 @@
+package ingestion
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go-sdk/access"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-evm-gateway/models"
+	errs "github.com/onflow/flow-evm-gateway/models/errors"
+	"github.com/onflow/flow-evm-gateway/services/requester"
+	"github.com/onflow/flow-evm-gateway/storage/pebble"
+)
+
+var _ models.Engine = &SealingVerifier{}
+
+// SealingVerifier verifies that soft finality events received over the Access polling API match the
+// actually sealed results from the event stream.
+type SealingVerifier struct {
+	*models.EngineStatus
+
+	logger zerolog.Logger
+
+	client *requester.CrossSporkClient
+	chain  flowGo.ChainID
+
+	eventsHash     *pebble.EventsHash
+	blocksToVerify map[uint64]flow.Identifier
+
+	mu sync.Mutex
+}
+
+// NewSealingVerifier creates a new sealing verifier.
+func NewSealingVerifier(
+	logger zerolog.Logger,
+	client *requester.CrossSporkClient,
+	chain flowGo.ChainID,
+	eventsHash *pebble.EventsHash,
+) *SealingVerifier {
+	return &SealingVerifier{
+		EngineStatus: models.NewEngineStatus(),
+		logger:       logger,
+		client:       client,
+		chain:        chain,
+		eventsHash:   eventsHash,
+	}
+}
+
+// Stop the engine.
+func (v *SealingVerifier) Stop() {
+	v.MarkDone()
+	<-v.Stopped()
```

Comment:
the done and stopped functionality is confusing. I modeled what the ingestion engine does.